README.md
Welcome to WebGraph (big)!
--------------------------

[WebGraph](http://webgraph.di.unimi.it/) is a framework for graph
compression aimed at studying web graphs. It provides simple ways to
manage very large graphs, exploiting modern compression techniques.

This big version of WebGraph can manage graphs as large as your core memory
allows. For graphs with less than 2^31 nodes, have a look at the [standard
version](https://github.com/vigna/webgraph).

Building
--------

You need [Ant](https://ant.apache.org/) and [Ivy](https://ant.apache.org/ivy/).
Then, run `ant ivy-setupjars jar`.

seba (<mailto:sebastiano.vigna@unimi.it>)


CHANGES
3.7.1

- Fixed old bug in UnionArcLabelledImmutableGraph.

- BitStreamArcLabelledImmutableGraph has now copiable iterators.

3.7.0

- New BidirectionalImmutableGraph wrapper to manage easily a graph
  and its transpose. Thanks to Antoine Pietri for contributing this
  class.

- New FixedWidthLongListLabel class. Thanks to Antoine Pietri and
  Thibault Allançon for contributing this class.

- Instances of EliasFanoMonotoneLongBigList16 have been replaced with
  instances of EliasFanoMonotoneBigLongBigList. As a side-effect,
  the number of bits per pointer will not be written unless you 
  recompute your .obl files.

- BitStreamArcLabelledImmutableGraph now supports Elias-Fano storage
  of label offsets. Thanks to Antoine Pietri for implementing this
  feature.

3.6.6

- Now ScatteredArcsASCIIGraph propagates BatchGraph's copiable iterators.
  As a consequence, compression will happen in parallel when possible.

- Now ScattereArcsASCIIGraph accepts functions on byte arrays.

- Official support starts now at Java 9.

3.6.5

- Updated licensing info in the POM.

3.6.4

- Removed (almost) unused dependencies.

- Upgraded to newest versions of Apache Commons.

- The command-line options of Transform now are the same as those
  of the standard version.

3.6.3

- OSGi metadata and default modularization.

- Manual reduction in strength in arithmetic operations everywhere.

3.6.2

- WebGraph (big) is now dually licensed under the Lesser GNU Public
  License 2.1+ or the Apache Software License 2.0.

- Added automatic module name.

3.6.1

- Ported ImmutableSubgraph from the standard version (thanks to
  Antoine Pietri for providing the port).

- New skip() method with long argument in NodeIterator.

3.6.0

- Ported from the standard version parallel compression based on the
  notion of "copiable iterators". Implemented in BVGraph and all
  derivative classes (e.g., transposed graphs).

- Transform provides direct support to compute efficiently the simple version
  of a graph, either directly or via a precomputed transpose.

3.5.3

- Fixed SLF4J dependencies (again).

3.5.2

- Fixed build.xml: moved to Java 9 (following Sux4J), and now use the
  classifier in naming artifacts.

3.5.1

- ScatteredArcsASCIIGraph was checking size() instead of trying
  size64() on the provided function. Thanks to Thibault Allançon
  for reporting this bug.

- ScatteredArcsASCIIGraph can now accept zipped files. Thanks to Thibault
  Allançon for implementing this feature.

- BVGraph will use EliasFanoMonotoneLongBigList16 when necessary.

3.5.0

- Java 8-only.

- Fixed obscure bug in ShiftByOneArcListASCIIGraph: if the arc list
  was specified on the command line (no -1 option) and more than
  one core was available, the graph would have not been shifted.
  Thanks to Luca Prigioniero for reporting this bug.

3.3.6

- The family of loadSequential() methods have been deprecated, and
  replaced in code by loadOffline() or loadMapped().

3.3.5

- Fixed dependencies.

3.3.4

- Significantly improved performance of HyperBall on graphs with a highly
  skewed (e.g., heavy-tailed) outdegree distribution (e.g., transposed web
  graphs).

- Fixed wrong estimation of memory used.

- Now ConnectedComponents writes results using "wcc" instead of "scc".

3.3.3

- Regressed to fastutil's quicksort calls in case of array fragments. Java
  7's Arrays.sort() has a memory bug that was killing the performance of a
  number of methods.

3.3.2

- We now distribute SpeedTest, hoping to improve the quality of benchmarks
  in the literature.

3.3.1

- Adapted to new DSI utilities.

3.3.0

- HyperBall sports a new adaptive decomposition scheme that 
  is based on the number of arcs to be scanned, rather than
  on the number of nodes.

- Fixed bug in the computation of the buckets. If you have used the new
  iterative implementation of Tarjan's algorithm
  (StronglyConnectedComponents) to compute buckets please recompute them.

- ParallelBreadthFirstVisit and ConnectedComponents have been ported
  from the standard version.

3.2.1

- New iterative implementation of Tarjan's algorithm.

- HyperBall can now compute Nieminen's centrality.

- Added missing shift option to ArcListASCIIGraph.

3.2.0

- New selectable upper bound for EFGraph makes it possible to build
  "fake" graphs in which successors are greater than or equal to
  the number of nodes (this was already possible with BVGraph). Useful
  for incremental graph construction.

- New IncrementalImmutableSequentialGraph adapter, which provides an
  inversion of control for storing graphs: you supply, one at a time,
  the successor list of each node.

3.1.0

- We switched to SLF4J for logging.

- Now ScatteredArcsASCIIGraph accepts a translation function from
  node identifiers to node numbers.

- The DecimalFormat object used to print data is has now a fixed US locale.

- New EFGraph implementation using the Elias-Fano representation of
  monotone sequences. Compression is not so good, but successor
  enumeration is blazingly fast and the implementation returns a skippable
  iterator which provides constant-time search of nodes by lower bound.

- Both BVGraph and EFGraph have outdegree caching and exact unwrapping
  of successorArray(). This should bring performance improvements.

- New HyperBall implementation of the HyperANF idea ported to this
  version. It computes several kind of geometric centrality and once in
  systolic local mode uses time proportional to the number of edges
  causing a modification, setting in practice the expected run time to the
  theoretical bound O(m log n).

- Several wrong instances of "int" have been replaced with "long".

3.0.3

- New ImmutableGraph.outdegrees() method that exposes the outdegrees of a
  graph as a LongIterator.

- RandomGraph removed.

- Almost all transformed graphs now support copy().

- New Transform.NodeClassFilter.

3.0.1

- ASCIIGraph was parsing the number of nodes as an integer rather than a

src/it/unimi/dsi/big/webgraph/Stats.java
/*
 * Copyright (C) 2007-2022 Sebastiano Vigna
 *
 * This program and the accompanying materials are made available under the
 * terms of the GNU Lesser General Public License v2.1 or later,
 * which is available at
 * http://www.gnu.org/licenses/old-licenses/lgpl-2.1-standalone.html,
 * or the Apache Software License 2.0, which is available at
 * https://www.apache.org/licenses/LICENSE-2.0.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.
 *
 * SPDX-License-Identifier: LGPL-2.1-or-later OR Apache-2.0
 */

package it.unimi.dsi.big.webgraph;

import static it.unimi.dsi.fastutil.BigArrays.get;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.lang.reflect.InvocationTargetException;
import java.math.BigDecimal;
import java.math.BigInteger;
import java.math.RoundingMode;
import java.util.Arrays;

import com.martiansoftware.jsap.FlaggedOption;
import com.martiansoftware.jsap.JSAP;
import com.martiansoftware.jsap.JSAPException;
import com.martiansoftware.jsap.JSAPResult;
import com.martiansoftware.jsap.Parameter;
import com.martiansoftware.jsap.SimpleJSAP;
import com.martiansoftware.jsap.Switch;
import com.martiansoftware.jsap.UnflaggedOption;

import it.unimi.dsi.big.webgraph.algo.StronglyConnectedComponents;
import it.unimi.dsi.bits.Fast;
import it.unimi.dsi.bits.LongArrayBitVector;
import it.unimi.dsi.fastutil.BigArrays;
import it.unimi.dsi.fastutil.io.BinIO;
import it.unimi.dsi.fastutil.io.TextIO;
import it.unimi.dsi.fastutil.longs.LongArrays;
import it.unimi.dsi.fastutil.longs.LongBigArrays;
import it.unimi.dsi.logging.ProgressLogger;

/** Computes basic statistical data about a given graph.
 *
 * <p>This class loads a graph of given basename, and computes the following data:
 * <ol>
 * <li>an ASCII file containing the <em>outdegree distribution</em>; line <var>n</var> contains the number of nodes with outdegree <var>n</var> (starting from 0);
 * <li>an ASCII file containing the <em>indegree distribution</em>; line <var>n</var> contains the number of nodes with indegree <var>n</var> (starting from 0);
 * <li>a property file containing several self-descriptive data, such as the average indegree/outdegree (which should be identical), sample nodes with minimum
 * or maximum indegree/outdegree, and so on; additional data will be computed if files produced by {@link StronglyConnectedComponents} are present
 * with the same basename (in particular, buckets and component sizes);
 * <li>if files produced by {@link StronglyConnectedComponents} are present with the same basename, an ASCII file containing the <em>distribution
 * of strongly connected components</em>, specified as a sequence of lines each containing a pair of integer &lt;<var>size</var>, <var>count</var>&gt;.
 * </ol>
 *
 * <p>The graph is loaded {@linkplain ImmutableGraph#loadOffline(CharSequence) offline}: the only memory allocated is for indegree count (one integer
 * per node) and for storing the actual counts (one integer per indegree/outdegree value).
 */

public class Stats {

	private Stats() {}

	/** Computes stats for the given graph using a single traversal, storing the results in files with given basename.
	 *
	 * @param graph the graph to be examined.
	 * @param buckets the set of buckets of this graph, or <code>null</code> if this information is not available.
	 * @param sccsize the sizes of strongly connected components as a big array, or <code>null</code> if this information is not available.
	 * @param resultsBasename the basename for result files (see the {@linkplain Stats class description}).
	 * @param pl a progress logger.
	 */

	public static void run(final ImmutableGraph graph, final LongArrayBitVector buckets, final long[][] sccsize, final CharSequence resultsBasename, final ProgressLogger pl) throws IOException {
		run(graph, buckets, sccsize, resultsBasename, false, pl);
	}

	/** Computes stats for the given graph using a single traversal, storing the results in files with given basename.
	 *
	 * @param graph the graph to be examined.
	 * @param buckets the set of buckets of this graph, or <code>null</code> if this information is not available.
	 * @param sccsize the sizes of strongly connected components as a big array, or <code>null</code> if this information is not available.
	 * @param resultsBasename the basename for result files (see the {@linkplain Stats class description}).
	 * @param saveDegrees if true, indegrees and outdegrees will be saved.
	 * @param pl a progress logger.
	 */

	public static void run(final ImmutableGraph graph, final LongArrayBitVector buckets, final long[][] sccsize, final CharSequence resultsBasename, final boolean saveDegrees, final ProgressLogger pl) throws IOException {
		final NodeIterator nodeIterator = graph.nodeIterator();
		long[] count = LongArrays.EMPTY_ARRAY;
		long[][] successor;
		final long[][] indegree = LongBigArrays.newBigArray(graph.numNodes());
		long curr, d, maxd = 0, maxNode = 0, mind = Long.MAX_VALUE, minNode = 0, dangling = 0, terminal = 0, loops = 0;
		long numArcs = 0, numGaps = 0;
		BigInteger totLoc = BigInteger.ZERO, totGap = BigInteger.ZERO;


		if (pl != null) {
			pl.itemsName = "nodes";
			pl.expectedUpdates = graph.numNodes();
			pl.start("Scanning...");
		}

		final PrintWriter outdegreesPrintWriter = saveDegrees ? new PrintWriter(new BufferedWriter(new FileWriter(resultsBasename + ".outdegrees"))) : null;

		/** Statistics for the gap width of successor lists (exponentially binned). */
		final long[] successorDeltaStats = new long[64];

		for(long i = graph.numNodes(); i-- != 0;) {
			curr = nodeIterator.nextLong();
			d = nodeIterator.outdegree();
			if (d > Integer.MAX_VALUE) throw new IllegalArgumentException("The current implementation of " + Stats.class.getSimpleName() + " cannot handle outdegrees beyond 2^31.");
			if (saveDegrees) outdegreesPrintWriter.println(d);
			successor = nodeIterator.successorBigArray();

			if (d > 1) {
				totGap = totGap.add(BigInteger.valueOf(get(successor, d - 1) - get(successor, 0)));
				totGap = totGap.add(BigInteger.valueOf(Fast.int2nat(get(successor, 0) - curr)));
				numGaps += d ;
			}
			for(long s = d; s-- != 0;) {
				totLoc = totLoc.add(BigInteger.valueOf(Math.abs(get(successor, s) - curr)));

				if (get(successor, s) != curr) successorDeltaStats[Fast.mostSignificantBit(Math.abs(curr - get(successor, s)))]++;
				else loops++;

				BigArrays.incr(indegree, get(successor, s));
			}

			if (d == 0) {
				dangling++;
				terminal++;
			}

			if (d == 1 && get(successor, 0) == curr) terminal++;

			if (d < mind) {
				mind = d;
				minNode = curr;
			}

			if (d > maxd){
				maxd = d;
				maxNode = curr;
			}

			numArcs += d;

			if (d >= count.length) count = LongArrays.grow(count, (int)d + 1);
			count[(int)d]++;

			if (pl != null) pl.lightUpdate();
		}

		if (pl != null) pl.done();

		if (saveDegrees) {
			outdegreesPrintWriter.close();
			TextIO.storeLongs(indegree, resultsBasename + ".indegrees");
		}

		@SuppressWarnings("resource")
		final
		PrintWriter properties = new PrintWriter(new FileWriter(resultsBasename + ".stats"));
		properties.println("nodes=" + graph.numNodes());
		properties.println("arcs=" + numArcs);
		properties.println("loops=" + loops);
		properties.println("successoravggap=" + new BigDecimal(totGap).divide(BigDecimal.valueOf(Math.max(1, numGaps)), 3, RoundingMode.HALF_EVEN));
		properties.println("avglocality=" + new BigDecimal(totLoc).divide(BigDecimal.valueOf(Math.max(1, numArcs)), 3, RoundingMode.HALF_EVEN));
		properties.println("minoutdegree=" + mind);
		properties.println("maxoutdegree=" + maxd);
		properties.println("minoutdegreenode=" + minNode);
		properties.println("maxoutdegreenode=" + maxNode);
		properties.println("dangling=" + dangling);
		properties.println("terminal=" + terminal);
		properties.println("percdangling=" + 100.0 * dangling / graph.numNodes());
		properties.println("avgoutdegree=" + (double)numArcs/graph.numNodes());

		int l;
		for(l = successorDeltaStats.length; l-- != 0;) if (successorDeltaStats[l] != 0) break;
		final StringBuilder s = new StringBuilder();
		double totLogDelta = 0;
		long numDelta = 0;

		long g = 1;
		for(int i = 0; i <= l; i++) {
			if (i != 0) s.append(',');
			s.append(successorDeltaStats[i]);
			numDelta += successorDeltaStats[i];
			totLogDelta += (Fast.log2(g * 2 + g + 1) - 1) * successorDeltaStats[i];
			g *= 2;
		}

		properties.println("successorlogdeltastats=" + s.toString());
		properties.println("successoravglogdelta=" + (numDelta == 0 ? "0" : new BigDecimal(totLogDelta).divide(BigDecimal.valueOf(Math.max(1, numDelta * 2)), 3, RoundingMode.HALF_EVEN).toString()));

		TextIO.storeLongs(count, 0, (int)(maxd + 1), resultsBasename + ".outdegree");

		Arrays.fill(count, 0);

		maxd = maxNode = minNode = 0;
		mind = Long.MAX_VALUE;
		long n = graph.numNodes();
		for(int i = indegree.length; i-- != 0;) {
			final long[] t = indegree[i];
			for(int p = t.length; p-- != 0;) {
				n--;
				d = t[p];
				if (d > Integer.MAX_VALUE) throw new IllegalArgumentException("The current implementation of " + Stats.class.getSimpleName() + " cannot handle indegrees beyond 2^31.");
				if (d >= count.length) count = LongArrays.grow(count, (int)(d + 1));
				if (d < mind) {
					mind = d;
					minNode = n;
				}

				if (d > maxd){
					maxd = d;
					maxNode = n;
				}

				count[(int)d]++;
			}
		}

		TextIO.storeLongs(count, 0, (int)(maxd + 1), resultsBasename + ".indegree");

		properties.println("minindegree=" + mind);
		properties.println("maxindegree=" + maxd);
		properties.println("minindegreenode=" + minNode);
		properties.println("maxindegreenode=" + maxNode);
		properties.println("avgindegree=" + (double)numArcs/graph.numNodes());

		if (buckets != null) {
			final long numBuckets = buckets.count();
			properties.println("buckets=" + numBuckets);
			properties.println("percbuckets=" + 100.0 * numBuckets / graph.numNodes());
		}

		if (sccsize != null) {
			LongBigArrays.quickSort(sccsize);
			final long m = BigArrays.length(sccsize);
			final long maxSize = get(sccsize, m - 1);
			final long minSize = get(sccsize, 0);

			properties.println("sccs=" + m);
			properties.println("maxsccsize=" + maxSize);
			properties.println("percmaxscc=" + 100.0 * maxSize / graph.numNodes());
			properties.println("minsccsize=" + minSize);
			properties.println("percminscc=" + 100.0 * minSize / graph.numNodes());

			final PrintWriter pw = new PrintWriter(resultsBasename + ".sccdistr");
			long current = maxSize;
			int c = 0;
			for(int i = sccsize.length; i-- != 0;) {
				final long[] t = sccsize[i];
				for(int j = t.length; j-- != 0;) {
					if(t[j] != current) {
						pw.println(current + "\t" + c);
						current = t[j];
						c = 0;
					}
					c++;
				}
			}

			pw.println(current + "\t" + c);

			pw.flush();
			pw.close();
		}

		properties.close();
	}

	static public void main(final String arg[]) throws IllegalArgumentException, SecurityException, IllegalAccessException, InvocationTargetException, NoSuchMethodException, JSAPException, IOException, ClassNotFoundException {
		final SimpleJSAP jsap = new SimpleJSAP(Stats.class.getName(), "Computes statistical data of a given graph.",
				new Parameter[] {
						new FlaggedOption("graphClass", GraphClassParser.getParser(), null, JSAP.NOT_REQUIRED, 'g', "graph-class", "Forces a Java class for the source graph."),
						new FlaggedOption("logInterval", JSAP.LONG_PARSER, Long.toString(ProgressLogger.DEFAULT_LOG_INTERVAL), JSAP.NOT_REQUIRED, 'l', "log-interval", "The minimum time interval between activity logs in milliseconds."),
						new Switch("saveDegrees", 's', "save-degrees", "Save indegrees and outdegrees in text format."),
						new UnflaggedOption("basename", JSAP.STRING_PARSER, JSAP.NO_DEFAULT, JSAP.REQUIRED, JSAP.NOT_GREEDY, "The basename of the graph."),
						new UnflaggedOption("resultsBasename", JSAP.STRING_PARSER, JSAP.NO_DEFAULT, JSAP.NOT_REQUIRED, JSAP.NOT_GREEDY, "The basename of the resulting files."),
					}
				);

		final JSAPResult jsapResult = jsap.parse(arg);
		if (jsap.messagePrinted()) System.exit(1);

		final Class<?> graphClass = jsapResult.getClass("graphClass");
		final String basename = jsapResult.getString("basename");
		final String resultsBasename = jsapResult.userSpecified("resultsBasename") ? jsapResult.getString("resultsBasename") : basename;

		final ProgressLogger pl = new ProgressLogger();
		pl.logInterval = jsapResult.getLong("logInterval");

		final ImmutableGraph graph;

		if (graphClass != null) graph = (ImmutableGraph)graphClass.getMethod("loadOffline", CharSequence.class).invoke(null, basename);
		else graph = ImmutableGraph.loadOffline(basename, pl);

		final LongArrayBitVector buckets = (LongArrayBitVector)(new File(basename + ".buckets").exists() ? BinIO.loadObject(basename + ".buckets") : null);
		final long[][] sccsize = new File(basename + ".sccsizes").exists() ? BinIO.loadLongsBig(basename + ".sccsizes") : null;

		run(graph, buckets, sccsize, resultsBasename, jsapResult.getBoolean("saveDegrees"), pl);
	}
}
